### PR TITLE
Laboratory work #1, Aisina Alina - 21FPL1

### PR DESCRIPTION
**Pull request by itsaisina**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -2,158 +2,256 @@
...
-
-    Returns:
-    list[str]: A sequence of lowercase tokens with no punctuation
-
-    In case of corrupt input arguments, None is returned
+    Checks that type of 'list1' is list and verifies the contents of 'list1' with the type that the user specifies
     """
-    pass
+    if isinstance(list1, list):
+        if list1 == [] and empty is False:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:15]: simplify the condition. You don't need to directly compare with `False` and an empty list. https://www.freecodecamp.org/news/truthy-and-falsy-values-in-python/

File: lab_1_keywords_tfidf/main.py [hunk 2]
```diff
@@ -2,158 +2,256 @@
...
+            return False
+        for index in list1:
+            if isinstance(index, type1) is False:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:18]: again

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -2,158 +2,256 @@
...
-    Returns:
-    List[str]: Token sequence that does not include stop words
-
-    In case of corrupt input arguments, None is returned
+    Checks that type of 'dictionary' is dict and verifies the contents of 'dictionary' with the type that
+    the user specifies
     """
-    pass
+    if isinstance(dictionary, dict):
+        if dictionary == {} and empty is False:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:30]: the same

File: lab_1_keywords_tfidf/main.py [hunk 4]
```diff
@@ -2,158 +2,256 @@
...
     """
-    Composes a frequency dictionary from the token sequence
+        Removes punctuation, casts to lowercase, splits into tokens
+        Parameters:
+        text (str): Original text
+        Returns:
+        list[str]: A sequence of lowercase tokens with no punctuation
+        In case of corrupt input arguments, None is returned
+        """
+    if not isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:48]: Use the same mechanism of early return in function above

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -2,158 +2,256 @@
...
+                return False
+        return True
+    return False
 
 
-def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
+def clean_and_tokenize(text: str) -> Optional[list[str]]:
     """
-    Composes a frequency dictionary from the token sequence
+        Removes punctuation, casts to lowercase, splits into tokens
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:41]: Why are docstrings changed? Please return them to their original condition

File: lab_1_keywords_tfidf/main.py (hunk 1 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 2 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 3 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 5 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py:41]: It's not my changes

File: seminars/practice_1_intro.py [hunk 6]
```diff
@@ -28,7 +28,7 @@
 # third_num = first_num / second_num
 
 # Fix the first exception:
-second_num = 6
+second_num = 5
```
Comment (artyomtugaryov) [inline seminars/practice_1_intro.py:31]: Please revert this change

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -2,158 +2,256 @@
...
-    text (str): Original text
-
-    Returns:
-    list[str]: A sequence of lowercase tokens with no punctuation
-
-    In case of corrupt input arguments, None is returned
+    Checks that type of 'list1' is list and verifies the contents of 'list1' with the type that the user specifies
     """
-    pass
+    if isinstance(list1, list):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:14]: try to check that `list1` is NOT a list and return `False` immediately.

File: lab_1_keywords_tfidf/main.py [hunk 8]
```diff
@@ -2,158 +2,256 @@
...
-
-    Returns:
-    List[str]: Token sequence that does not include stop words
-
-    In case of corrupt input arguments, None is returned
+    Checks that type of 'dictionary' is dict and verifies the contents of 'dictionary' with the type that
+    the user specifies
     """
-    pass
+    if isinstance(dictionary, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:29]: the same

File: lab_1_keywords_tfidf/main.py (hunk 5 above)
Comment (artyomtugaryov) [reply lab_1_keywords_tfidf/main.py:41]: But GitHub shows that you changed it. Anyway - return every docstring to the original view

File: lab_1_keywords_tfidf/main.py [hunk 9]
```diff
@@ -2,158 +2,256 @@
...
     """
-    Composes a frequency dictionary from the token sequence
+        Removes punctuation, casts to lowercase, splits into tokens
+        Parameters:
+        text (str): Original text
+        Returns:
+        list[str]: A sequence of lowercase tokens with no punctuation
+        In case of corrupt input arguments, None is returned
+        """
+    if not isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:48]: 👍

File: lab_1_keywords_tfidf/main.py [hunk 10]
```diff
@@ -2,158 +2,256 @@
...
+        return None
+    lowered_text = text.lower()
+    for punctuation_mark in string.punctuation:
+        if punctuation_mark in lowered_text:
+            lowered_text = lowered_text.replace(punctuation_mark, '')
+    cleaned_text = lowered_text.strip().split()
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:54]: You can return `lowered_text.strip().split()` directly. without creating the variable

File: lab_1_keywords_tfidf/main.py [hunk 11]
```diff
@@ -2,158 +2,256 @@
...
-    pass
+        Excludes stop words from the token sequence
+        Parameters:
+        tokens (List[str]): Original token sequence
+        stop_words (List[str]: Tokens to exclude
+        Returns:
+        List[str]: Token sequence that does not include stop words
+        In case of corrupt input arguments, None is returned
+        """
+    if correct_list(tokens, str, False) and correct_list(stop_words, str, True):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:68]: reverse the condition and return `None` as soon as possible

File: lab_1_keywords_tfidf/main.py [hunk 12]
```diff
@@ -2,158 +2,256 @@
...
-    its corresponding frequency values
-    top (int): Number of token to extract
+        Composes a frequency dictionary from the token sequence
+        Parameters:
+        tokens (List[str]): Token sequence to count frequencies for
+        Returns:
+        Dict: {token: number of occurrences in the token sequence} dictionary
+        In case of corrupt input arguments, None is returned
+        """
+    if correct_list(tokens, str, False):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:90]: again

File: lab_1_keywords_tfidf/main.py [hunk 13]
```diff
@@ -2,158 +2,256 @@
...
+        Parameters:
+        frequencies (Dict): A dictionary with tokens and
+        its corresponding frequency values
+        top (int): Number of token to extract
+        Returns:
+        List[str]: Sequence of specified length
+        consisting of tokens with the largest frequency
+        In case of corrupt input arguments, None is returned
+        """
+    if correct_dict(frequencies, str, float, False) and (not isinstance(top, bool) and isinstance(top, int)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:115]: again

File: lab_1_keywords_tfidf/main.py [hunk 14]
```diff
@@ -2,158 +2,256 @@
...
+        based on the raw frequency
+        Parameters:
+        frequencies (Dict): Raw number of occurrences for each of the tokens
+        Returns:
+        dict: A dictionary with tokens and corresponding term frequency score
+        In case of corrupt input arguments, None is returned
+        """
+    all_words = 0
+    new_dict = {}
+    if correct_dict(frequencies, str, int, False):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:139]: again - reverse the condition

File: lab_1_keywords_tfidf/main.py [hunk 15]
```diff
@@ -2,158 +2,256 @@
...
-    Dict: A dictionary with tokens and its corresponding TF-IDF values
+        Returns:
+        Dict: A dictionary with tokens and its corresponding TF-IDF values
 
-    In case of corrupt input arguments, None is returned
-    """
-    pass
+        In case of corrupt input arguments, None is returned
+        """
+    if correct_dict(term_freq, str, float, False) and correct_dict(idf, str, float, True):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:162]: the same

File: seminars/practice_1_intro.py (hunk 6 above)
Comment (itsaisina) [reply seminars/practice_1_intro.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 7 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 8 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 10 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 11 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 12 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 13 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 14 above)
Comment (itsaisina) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py [hunk 16]
```diff
@@ -48,7 +86,17 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
...
     """
-    pass
+    if not correct_list(tokens, str, False):
+        return None
+    frequency_dict = {}
+    for token in tokens:
+        if not isinstance(token, str):
+            return None
+        if token in frequency_dict.keys():
+            frequency_dict[token] = 1 + frequency_dict[token]
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:96]: use the `+=` operator

**Review by artyomtugaryov — APPROVED** [review]

File: lab_1_keywords_tfidf/main.py [hunk 17]
```diff
@@ -33,7 +66,12 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not correct_list(tokens, str, False) or not correct_list(stop_words, str, True):
+        return None
+    for stop_word in stop_words:
+        while stop_word in tokens:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:72]: .

File: lab_1_keywords_tfidf/main.py [hunk 18]
```diff
@@ -48,7 +86,15 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not correct_list(tokens, str, False):
+        return None
+    frequency_dict = {}
+    for token in tokens:
+        if token in frequency_dict.keys():
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:93]: dict.get()
